tcp/server.go

```diff
@@ -33,113 +25,82 @@ func NewServer(h Handler) (*Server, error) {
 	}
 
 	return &Server{
-		handler:   h,
-		listeners: map[*net.Listener]struct{}{},
+		handler: h,
 	}, nil
 }
 
 // Listen listens to an address for new connections, passing them
 // off to the handler in a goroutine.
-func (s *Server) Listen(addr string) error {
-	ln, err := net.Listen("tcp", addr)
+func (s *Server) Listen(ctx context.Context, addr string) error {
+	lc := &net.ListenConfig{}
+	ln, err := lc.Listen(ctx, "tcp", addr)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to listen: %w", err)
 	}
-
-	ln = &onceCloseListener{Listener: ln}
 	defer func() { _ = ln.Close() }()
 
-	if !s.addListener(&ln) {
-		return ErrServerClosed
-	}
-	defer s.removeListener(&ln)
+	go func() {
+		<-ctx.Done()
+		_ = ln.Close()
+	}()
 
+	var conn net.Conn
 	for {
-		conn, err := ln.Accept()
-		if err != nil {
-			if s.inShutdown.isSet() {
-				return ErrServerClosed
-			}
-
-			var ne net.Error
-			if errors.As(err, &ne) && ne.Timeout() {
-				continue
-			}
+		conn, err = ln.Accept()
 
+		var netErr net.Error
+		switch {
+		case err != nil && errors.Is(err, net.ErrClosed):
+			return err
+		case err != nil && errors.As(err, &netErr) && netErr.Timeout():
 			return err
+		case err != nil:
+			return fmt.Errorf("failed to accept connection: %w", err)
 		}
 
-		if s.inShutdown.isSet() {
-			return ErrServerClosed
-		}
+		go func() {
+			defer func() { _ = conn.Close() }()
 
-		go s.handler.ServeTCP(conn)
+			s.handler.Serve(&packetConn{conn})
+		}()
 	}
 }
 
-func (s *Server) addListener(ln *net.Listener) bool {
-	s.mu.Lock()
-	defer s.mu.Unlock()
-
-	if s.inShutdown.isSet() {
-		return false
-	}
+var _ net.PacketConn = &packetConn{}
 
-	s.listeners[ln] = struct{}{}
-	return true
+// packetConn makes a TCP connection act like an unbound connection
+// to support the same interface that a UDP connection offers.
+//
+// The other way round of making a UDP connection act like a bound connection just like TCP is more complex.
```

Comment:
I'm not sure I understand this sentence, what do you mean? 🤔

Reply:
I removed the sentence. It basically tells you why I have wrapped the TCP connection, and not the UDP connection (not even possbile AFAIK), or both.